### PR TITLE
Fix rebuild bug and add new subcommand

### PR DESCRIPTION
**Pull request by tiagolobocastro**
Check if there are any active tasks when completing the rebuild
Add new stats subcommand to retrieve the rebuild stats

### CODE REVIEW

**Comment by tiagolobocastro** [comment]
bors try

File: mayastor/src/rebuild/rebuild_impl.rs [hunk 1]
```diff
@@ -597,11 +599,21 @@ impl RebuildJob {
...
-        while self.await_one_task().await.is_some() {
-            if self.task_pool.active == 0 {
-                break;
+        debug!(
+            "Awaiting all active tasks({}) for rebuild {}",
+            self.task_pool.active, self.destination
+        );
+        if self.task_pool.active > 0 {
+            while self.await_one_task().await.is_some() {
+                if self.task_pool.active == 0 {
```
Comment (jkryl) [inline mayastor/src/rebuild/rebuild_impl.rs:608]: just a nit. The loop could be more compact:
```
while self.task_pool.active > 0 {
   self.await_one_task().await...
}
```

**Review by jkryl — APPROVED** [review]
looks good! We used to have a test for each CLI command. Those commands are using fake gRPC server so they don't test the real thing, but on the other hand it might be better to have at least some test compared to having none.

File: mayastor/src/rebuild/rebuild_impl.rs (hunk 1 above)
Comment (tiagolobocastro) [reply mayastor/src/rebuild/rebuild_impl.rs]: Thanks done, also added an error condition which although unlikely as we never close the channels.

**Comment by tiagolobocastro** [comment]
> looks good! We used to have a test for each CLI command. Those commands are using fake gRPC server so they don't test the real thing, but on the other hand it might be better to have at least some test compared to having none.

Good point, I'll try to add a test, do you know where the existing ones are?

File: mayastor/src/bdev/nexus/nexus_bdev_rebuild.rs [hunk 2]
```diff
@@ -324,3 +343,17 @@ impl Nexus {
         }
     }
 }
+
+impl From<RebuildStats> for RebuildStatsReply {
+    fn from(stats: RebuildStats) -> Self {
+        RebuildStatsReply {
+            blocks_total: stats.block_size,
```
Comment (paulyoong) [inline mayastor/src/bdev/nexus/nexus_bdev_rebuild.rs:350]: I don't think you meant stats.block_size for the total here.

File: mayastor/src/bin/cli/rebuild_cli.rs [hunk 3]
```diff
@@ -99,6 +100,21 @@ pub fn subcommands<'a, 'b>() -> App<'a, 'b> {
...
+            Arg::with_name("uuid")
+                .required(true)
+                .index(1)
+                .help("uuid of the nexus"),
+        )
+        .arg(
+            Arg::with_name("uri")
+                .required(true)
+                .index(2)
+                .help("uri of child to get the rebuild state from"),
```
Comment (paulyoong) [inline mayastor/src/bin/cli/rebuild_cli.rs:115]: "state" should be "stats"

File: rpc/proto/mayastor.proto [hunk 4]
```diff
@@ -292,6 +293,21 @@ message RebuildStateReply {
...
 
+message RebuildStatsRequest {
+  string uuid = 1;  // uuid of the nexus
+  string uri = 2;   // uri of the destination child
+}
+
+message RebuildStatsReply {
+  uint64 blocks_total = 1; // total number of blocks to recover
+  uint64 blocks_recovered = 2; // number of blocks recovered
+  uint64 progress = 3; // rebuild progress in % (0-100)
```
Comment (paulyoong) [inline rpc/proto/mayastor.proto:304]: nit - do you need "(0-100)" on the end? It's a % so it's implied

Comment (tiagolobocastro) [reply rpc/proto/mayastor.proto:304]: Done, also fixed the same comment from rebuild_api

**Review by paulyoong — APPROVED** [review]

**Comment by tiagolobocastro** [comment]
bors merge

**Comment by tiagolobocastro** [comment]
bors merge

**Comment by tiagolobocastro** [comment]
bors merge

**Comment by tiagolobocastro** [comment]
bors merge

**Pull request merged**